storage/postgres/notificator.go

```diff
@@ -103,73 +109,133 @@ func (n *Notificator) Start(ctx context.Context, group *sync.WaitGroup) error {
 	return nil
 }
 
-func (n *Notificator) RegisterConsumer(userContext *web.UserContext) (storage.NotificationQueue, int64, error) {
-	platform := &types.Platform{}
-	err := userContext.Data.Data(platform)
+func (n *Notificator) addConsumer(platform *types.Platform, queue storage.NotificationQueue) int64 {
+	n.consumersMutex.Lock()
+	defer n.consumersMutex.Unlock()
+	n.consumers.Add(platform, queue)
+	return atomic.LoadInt64(&n.lastKnownRevision)
+}
+
+func (n *Notificator) RegisterConsumer(platform *types.Platform, lastKnownRevision int64) (storage.NotificationQueue, int64, error) {
+	if atomic.LoadInt32(&n.isConnected) == aFalse {
+		return nil, types.INVALIDREVISION, errors.New("cannot register consumer - Notificator is not running")
+	}
+	queue, err := storage.NewNotificationQueue(n.queueSize)
 	if err != nil {
-		return nil, invalidRevisionNumber, fmt.Errorf("could not get platform from user context %v", err)
+		return nil, types.INVALIDREVISION, err
 	}
-	if platform.ID == "" {
-		return nil, invalidRevisionNumber, errors.New("platform ID not found in user context")
+	if err = n.startListening(); err != nil {
+		return nil, types.INVALIDREVISION, fmt.Errorf("listen to %s channel failed %v", postgresChannel, err)
 	}
-	if atomic.LoadInt32(&n.isConnected) == aFalse {
-		return nil, invalidRevisionNumber, errors.New("cannot register consumer - Notificator is not running")
+	lastKnownRevisionToSM := n.addConsumer(platform, queue)
+	if lastKnownRevision == types.INVALIDREVISION {
+		return queue, lastKnownRevisionToSM, nil
 	}
-	if err = n.startListening(); err != nil {
-		return nil, invalidRevisionNumber, fmt.Errorf("listen to %s channel failed %v", postgresChannel, err)
+	queueWithMissedNotifications, err := n.replaceQueueWithMissingNotificationsQueue(queue, lastKnownRevision, lastKnownRevisionToSM, platform)
+	if err != nil {
+		if errUnregisterConsumer := n.UnregisterConsumer(queue); errUnregisterConsumer != nil {
+			log.C(n.ctx).WithError(errUnregisterConsumer).Errorf("Could not unregister notification consumer %s", queue.ID())
+		}
+		return nil, types.INVALIDREVISION, err
 	}
-	queue, err := storage.NewNotificationQueue(n.queueSize)
+	return queueWithMissedNotifications, lastKnownRevisionToSM, nil
+}
+
+func (n *Notificator) filterRecipients(recipients []*types.Platform, notification *types.Notification) []*types.Platform {
+	for _, filter := range n.notificationFilters {
+		recipients = filter(recipients, notification)
+		if len(recipients) == 0 {
```

Comment:
filter returns an empty slice if any filter decides that no platform should receive it and the same list if all filters decide that at least one platform should receive it?

Reply:
This is an optimisation - if a filter in the chain of filters returns 0 recipients then it is not necessary to call the other filters as there are no platforms to filter. See also comment above.